File: 141-linkListCycle/solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        /*
        // 100%
        if (head == null || head.next==null) {
            return false;
        }
        ListNode slow = head, fast = head;
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
        */

        // 4%
        Set<ListNode> nodes = new HashSet<>();
        while(head!=null) {
            if (nodes.contains(head)) {
                return true;
            } else {
                nodes.add(head);    
                head=head.next;
            }
            
        }
        return false;
    }
}

File: 141-linkListCycle/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """

        # 24%
        if not(head) or not(head.next): return False
        
        fast, slow = head, head
        while (fast and fast.next):
            fast = fast.next.next
            slow = slow.next
            if fast==slow:
                return True
        return False
        
        # 60%
        nodes = set()
        while head:
            if head in nodes:
                return True
            else:
                nodes.add(head)
                head=head.next
                
        return False        

